File: app/src/main/java/com/fastaccess/ui/widgets/LabelSpan.java
package com.fastaccess.ui.widgets;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Paint.FontMetricsInt;
import android.graphics.RectF;
import android.support.annotation.NonNull;
import android.support.v4.text.TextUtilsCompat;
import android.support.v4.view.ViewCompat;
import android.text.Spanned;
import android.text.TextPaint;
import android.text.TextUtils;
import android.text.style.CharacterStyle;
import android.text.style.ReplacementSpan;

import com.fastaccess.helper.ViewHelper;

import java.util.Locale;


/**
 * adopted class from Android source code & modified to fit FastHub need.
 */

public class LabelSpan extends ReplacementSpan {
    public interface SpanDimensions {
        int getPadding();

        int getPaddingExtraWidth();

        int getPaddingAfter();

        int getMaxWidth();

        float getRoundedCornerRadius();

        int getMarginTop();

        boolean isRtl();
    }

    private final TextPaint txtPaint = new TextPaint();
    private final FontMetricsInt fontMetrics = new FontMetricsInt();
    private final int color;
    private final SpanDimensions dims;

    private LabelSpan(int color, @NonNull SpanDimensions dims) {
        this.color = color;
        txtPaint.bgColor = color;
        this.dims = dims;
    }

    public LabelSpan(int color) {
        this(color, new SpanDimensions() {
            @Override public int getPadding() {
                return 6;
            }

            @Override public int getPaddingExtraWidth() {
                return 0;
            }

            @Override public int getPaddingAfter() {
                return 0;
            }

            @Override public int getMaxWidth() {
                return 1000;//random number
            }

            @Override public float getRoundedCornerRadius() {
                return 5;
            }

            @Override public int getMarginTop() {
                return 8;
            }

            @Override public boolean isRtl() {
                return TextUtilsCompat.getLayoutDirectionFromLocale(Locale.getDefault()) == ViewCompat.LAYOUT_DIRECTION_RTL;
            }
        });
    }

    @Override public int getSize(@NonNull Paint paint, CharSequence text, int start, int end, FontMetricsInt fm) {
        setupFontMetrics(text, start, end, fm, paint);
        if (fm != null) {
            final int padding = dims.getPadding();
            final int margin = dims.getMarginTop();
            fm.ascent = Math.min(fm.top, fm.ascent - padding) - margin;
            fm.descent = Math.max(fm.bottom, padding);
            fm.top = fm.ascent;
            fm.bottom = fm.descent;
        }
        return measureWidth(txtPaint, text, start, end, dims.isRtl());
    }

    private int measureWidth(Paint paint, CharSequence text, int start, int end,
                             boolean includePaddingAfter) {
        final int paddingW = dims.getPadding() + dims.getPaddingExtraWidth();
        final int maxWidth = dims.getMaxWidth();
        int w = (int) paint.measureText(text, start, end) + 2 * paddingW;
        if (includePaddingAfter) {
            w += dims.getPaddingAfter();
        }
        if (w > maxWidth) {
            w = maxWidth;
        }
        return w;
    }

    private void setupFontMetrics(CharSequence text, int start, int end, FontMetricsInt fm, Paint p) {
        txtPaint.set(p);
        final CharacterStyle[] otherSpans = ((Spanned) text).getSpans(start, end, CharacterStyle.class);
        for (CharacterStyle otherSpan : otherSpans) {
            otherSpan.updateDrawState(txtPaint);
        }
        txtPaint.setTextSize(p.getTextSize());
        if (fm != null) {
            txtPaint.getFontMetricsInt(fm);
        }
    }

    @Override public void draw(@NonNull Canvas canvas, CharSequence text, int start, int end,
                               float x, int top, int y, int bottom, @NonNull Paint paint) {
        final int padding = dims.getPadding();
        final int paddingW = padding + dims.getPaddingExtraWidth();
        final int maxWidth = dims.getMaxWidth();
        setupFontMetrics(text, start, end, fontMetrics, paint);
        final int bgWidth = measureWidth(txtPaint, text, start, end, false);
        fontMetrics.top = Math.min(fontMetrics.top, fontMetrics.ascent - padding);
        fontMetrics.bottom = Math.max(fontMetrics.bottom, padding);
        top = y + fontMetrics.top - fontMetrics.bottom;
        bottom = y;
        y = bottom - fontMetrics.bottom;
        final boolean isRtl = dims.isRtl();
        final int paddingAfter = dims.getPaddingAfter();
        if (txtPaint.bgColor != 0) {
            final int prevColor = txtPaint.getColor();
            final Paint.Style prevStyle = txtPaint.getStyle();
            txtPaint.setColor(txtPaint.bgColor);
            txtPaint.setStyle(Paint.Style.FILL);
            final float left;
            if (isRtl) {
                left = x + paddingAfter;
            } else {
                left = x;
            }
            final float right = left + bgWidth;
            final RectF rect = new RectF(left, top, right, bottom);
            final float radius = dims.getRoundedCornerRadius();
            canvas.drawRoundRect(rect, radius, radius, txtPaint);
            txtPaint.setColor(prevColor);
            txtPaint.setStyle(prevStyle);
        }
        if (bgWidth == maxWidth) {
            text = TextUtils.ellipsize(text.subSequence(start, end).toString(), txtPaint, bgWidth - 2 * paddingW, TextUtils.TruncateAt.MIDDLE);
            start = 0;
            end = text.length();
        }
        float textX = x + paddingW;
        if (isRtl) {
            textX += paddingAfter;
        }
        if (color != Color.TRANSPARENT) txtPaint.setColor(ViewHelper.generateTextColor(color));
        canvas.drawText(text, start, end, textX, y, txtPaint);
    }
}

File: markdown/src/main/java/com/fastaccess/markdown/spans/LabelSpan.kt
package com.fastaccess.markdown.spans

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Paint.FontMetricsInt
import android.graphics.RectF
import android.text.Spanned
import android.text.TextPaint
import android.text.TextUtils
import android.text.style.CharacterStyle
import android.text.style.ReplacementSpan
import androidx.core.text.TextUtilsCompat
import androidx.core.view.ViewCompat
import com.fastaccess.github.extensions.generateTextColor
import java.util.*

/**
 * adopted class from Android source code & modified to fit FastHub need.
 */

class LabelSpan private constructor(private val color: Int, private val dims: SpanDimensions) : ReplacementSpan() {

    private val txtPaint = TextPaint()
    private val fontMetrics = FontMetricsInt()

    interface SpanDimensions {
        val padding: Int

        val paddingExtraWidth: Int

        val paddingAfter: Int

        val maxWidth: Int

        val roundedCornerRadius: Float

        val marginTop: Int

        val isRtl: Boolean
    }

    init {
        txtPaint.bgColor = color
    }

    constructor(color: Int, roundedCoroner: Float = 5f) : this(color, object : SpanDimensions {
        override val padding: Int
            get() = 6

        override val paddingExtraWidth: Int
            get() = 0

        override val paddingAfter: Int
            get() = 0

        override//random number
        val maxWidth: Int
            get() = 1000

        override val roundedCornerRadius: Float
            get() = roundedCoroner

        override val marginTop: Int
            get() = 8

        override val isRtl: Boolean
            get() = TextUtilsCompat.getLayoutDirectionFromLocale(Locale.getDefault()) == ViewCompat.LAYOUT_DIRECTION_RTL
    })

    override fun getSize(paint: Paint, text: CharSequence, start: Int, end: Int, fm: FontMetricsInt?): Int {
        setupFontMetrics(text, start, end, fm, paint)
        if (fm != null) {
            val padding = dims.padding
            val margin = dims.marginTop
            fm.ascent = Math.min(fm.top, fm.ascent - padding) - margin
            fm.descent = Math.max(fm.bottom, padding)
            fm.top = fm.ascent
            fm.bottom = fm.descent
        }
        return measureWidth(txtPaint, text, start, end, dims.isRtl)
    }

    private fun measureWidth(paint: Paint, text: CharSequence, start: Int, end: Int,
                             includePaddingAfter: Boolean): Int {
        val paddingW = dims.padding + dims.paddingExtraWidth
        val maxWidth = dims.maxWidth
        var w = paint.measureText(text, start, end).toInt() + 2 * paddingW
        if (includePaddingAfter) {
            w += dims.paddingAfter
        }
        if (w > maxWidth) {
            w = maxWidth
        }
        return w
    }

    private fun setupFontMetrics(text: CharSequence, start: Int, end: Int, fm: FontMetricsInt?, p: Paint) {
        txtPaint.set(p)
        val otherSpans = (text as Spanned).getSpans(start, end, CharacterStyle::class.java)
        for (otherSpan in otherSpans) {
            otherSpan.updateDrawState(txtPaint)
        }
        txtPaint.textSize = p.textSize
        if (fm != null) {
            txtPaint.getFontMetricsInt(fm)
        }
    }

    override fun draw(canvas: Canvas, text: CharSequence, start: Int, end: Int,
                      x: Float, top: Int, y: Int, bottom: Int, paint: Paint) {
        var textVal = text
        var startVal = start
        var endVal = end
        var yVal = y
        val padding = dims.padding
        val paddingW = padding + dims.paddingExtraWidth
        val maxWidth = dims.maxWidth
        setupFontMetrics(textVal, startVal, endVal, fontMetrics, paint)
        val bgWidth = measureWidth(txtPaint, textVal, startVal, endVal, false)
        fontMetrics.top = Math.min(fontMetrics.top, fontMetrics.ascent - padding)
        fontMetrics.bottom = Math.max(fontMetrics.bottom, padding)
        val topVal = yVal + fontMetrics.top - fontMetrics.bottom
        val bottomVal = yVal
        yVal = bottomVal - fontMetrics.bottom
        val isRtl = dims.isRtl
        val paddingAfter = dims.paddingAfter
        if (txtPaint.bgColor != 0) {
            val prevColor = txtPaint.color
            val prevStyle = txtPaint.style
            txtPaint.color = txtPaint.bgColor
            txtPaint.style = Paint.Style.FILL
            val left: Float = if (isRtl) {
                x + paddingAfter
            } else {
                x
            }
            val right = left + bgWidth
            val rect = RectF(left, topVal.toFloat(), right, bottomVal.toFloat())
            val radius = dims.roundedCornerRadius
            canvas.drawRoundRect(rect, radius, radius, txtPaint)
            txtPaint.color = prevColor
            txtPaint.style = prevStyle
        }
        if (bgWidth == maxWidth) {
            textVal = TextUtils.ellipsize(textVal.subSequence(startVal, endVal).toString(), txtPaint, (bgWidth - 2 * paddingW).toFloat(), TextUtils.TruncateAt.MIDDLE)
            startVal = 0
            endVal = textVal.length
        }
        var textX = x + paddingW
        if (isRtl) {
            textX += paddingAfter.toFloat()
        }
        if (color != Color.TRANSPARENT) txtPaint.color = txtPaint.color.generateTextColor()
        canvas.drawText(textVal, startVal, endVal, textX, yVal.toFloat(), txtPaint)
    }
}